lib/Conversion/CalyxNative/CalyxNative.cpp

```diff
@@ -0,0 +1,148 @@
+  if (!exeMb) {
+    root.emitError("cannot find executable: '" + execName + "'");
+    return failure();
+  }
+  StringRef calyxExe = exeMb.get();
+
+  std::string errMsg;
+  SmallString<32> nativeInputFileName;
+  std::error_code errCode = llvm::sys::fs::getPotentiallyUniqueTempFileName(
+      "calyxNativeTemp", /*suffix=*/"", nativeInputFileName);
+
+  if (std::error_code ok; errCode != ok) {
+    root.emitError(
+        "cannot generate a unique temporary file for input to Calyx compiler");
+    return failure();
+  }
+
+  // Emit the current program into a file so the native compiler can operate
+  // over it.
+  std::unique_ptr<llvm::ToolOutputFile> inputFile =
+      mlir::openOutputFile(nativeInputFileName, &errMsg);
+  if (inputFile == nullptr) {
+    root.emitError(errMsg);
+    return failure();
+  }
+
+  auto res = circt::calyx::exportCalyx(root, inputFile->os());
+  if (failed(res)) {
+    return failure();
+  }
+  inputFile->os().flush();
+
+  // Create a file for the native compiler to write the results into
+  SmallString<32> nativeOutputFileName;
+  errCode = llvm::sys::fs::getPotentiallyUniqueTempFileName(
+      "calyxNativeOutTemp", /*suffix=*/"", nativeOutputFileName);
+  if (std::error_code ok; errCode != ok) {
+    root.emitError(
+        "cannot generate a unique temporary file name to store output");
+    return failure();
+  }
+
+  std::optional<StringRef> redirects[] = {/*stdin=*/std::nullopt,
+                                          /*stdout=*/nativeOutputFileName,
+                                          /*stderr=*/std::nullopt};
+
+  auto args = llvm::ArrayRef<StringRef>{
+      calyxExe, nativeInputFileName, "-o", nativeOutputFileName,
+      "-l",     primitiveLib,        "-b", "mlir"};
+  int result = llvm::sys::ExecuteAndWait(calyxExe, args, /*Env=*/std::nullopt,
+                                         /*Redirects=*/redirects,
+                                         /*SecondsToWait=*/0, /*MemoryLimit=*/0,
+                                         &errMsg);
+
+  if (result != 0) {
+    root.emitError() << errMsg;
+    return failure();
+  }
+
+  // Parse the output buffer into a Calyx operation so that we can insert it
+  // back into the program.
+  auto bufferRead = llvm::MemoryBuffer::getFile(nativeInputFileName);
+  if (!bufferRead || !*bufferRead) {
+    root.emitError("execution of '" + calyxExe +
+                   "' did not produce any output file named '" +
+                   nativeInputFileName + "'");
+    return failure();
+  }
+
+  // Load the output from the native compiler as a ModuleOp
+  auto loadedMod =
+      parseSourceFile<ModuleOp>(nativeOutputFileName.str(), root.getContext());
+  auto *loadedBlock = loadedMod->getBody();
+
+  // XXX(rachitnigam): This is quite baroque. We insert the new block before the
+  // previous one and then remove the old block. A better thing to do would be
+  // to replace the moduleOp completely but I couldn't figure out how to do
+  // that.
```

Comment:
@cgyurgyik do you know of a better way to do this? Essentially, the Calyx compiler will generate a `module` for us so the best thing to do here would be to just replace `root` with `loadedMod`. However, after hours of looking, I was not able to find the right way to do this and settled on this hacky thing where I unlike the `block` from `loadedMod` and place it into `root` and then remove the original block in `root`.

Reply:
I think you would need to use a pattern rewriter for this to have access to rewriter functions. I'm pretty sure SCFToCalyx would have examples of this since we have rewrites there.

https://mlir.llvm.org/doxygen/classmlir_1_1RewriterBase.html